WIP: Create a rules.json file that can be used in both the frontend and backend.

seanlip opened the pull request:
/cc @BenHenning, @wxyxinyu, @stephenchiang (and @maitbayev, since I think he's interested)

This is an attempt to create a `rules.json` file that can be used by both the frontend and the backend. In the process, I have removed the entirety of `extensions/rules`, as well as most of the backend rule infrastructure.

Note that the tests do not currently pass. In order for this to be merged into develop, the following things still need to be done:
- [x] Migrate all rule evaluation to the frontend (and find a way to initiate calls from there in order to evaluate soft and string-classifier rules).
- [x] Figure out a good way to replace self.submit_answer() in `core.controllers.editor_test.EditorTest.test_untrained_answers_handler`, or possibly rewrite the test in some way.
- [x] Port the tests in `core.controllers.reader_test.ReaderClassifyTests.test_[hard|soft]_rule_classification` to the frontend.
- [x] Verify that the e2e tests pass.

The crux of all these issues is the `_evaluate_rule()` function in `core/controllers/reader.py`, which is problematic because we no longer evaluate rules on the backend. Most of the logic here, as well as that in the three `classify_*_rule` functions, should probably move to the frontend (or be invoked as a stateless call from the frontend).

One thing I do not understand (and would appreciate clarification by @BenHenning / @stephenchiang) is the point of soft rules. I've been out of the loop here for a bit, but I thought that we have two types of rules: hard and fuzzy (where the latter category includes string-classifier). It seems now that we have hard, soft, and string-classifier. I wanted to ask what an example of a viable 'soft' rule is, since I believe all the current 'soft' rules we have are, to some extent, placeholders. If we don't have any viable ones, I would be inclined to suggest that we only have two categories of rules instead, since that would simplify things somewhat.

In addition, another thing I don't understand is why the string-classifier rule is a top-level function in reader.py, whereas I would have expected it to be essentially the 'soft' rule for TextInput. Are we planning to use it for other interactions as well?

Finally, @wxyxinyu, would you mind taking this over when you get back? I suspect I won't be able to look at it in the near future, but hopefully the work I've done so far takes us 70% of the way or so towards getting rid of rules as a separate construct.

Thanks!

seanlip commented:
Also, if time permits, I would appreciate reviews from any two of @BenHenning, @wxyxinyu and @stephenchiang on what I've done so far.

maitbayev commented on extensions/interactions/rules.json:
```diff
@@ -0,0 +1,186 @@
+{
+  "CodeRepl": {
```
can extensions/interactions/CodeRepl contain rules.json? The same for other interactions, so that they are self-contained?

seanlip replied:
It's a good question -- I'm a bit torn. On the one hand it's nice that they are self-contained within their interactions, and I can see the attraction of that. On the other hand, it's simpler and cheaper for the backend/frontend to read from a single file -- kind of like feconf.py. (Ultimately we'll probably refer to this file from the frontend as well, for getting the rule definitions.) It's also easier for the Karma/Protractor tests to reference a single rules.json, since it's a bit trickier to run a for loop over json imports.

For now, I lean towards keeping everything in a single file, and writing Karma tests to ensure that the rule evaluation definitions for each interaction coincide with what's in this file (which is the approach I've taken in this commit), so that any skew between the specifications will be quickly noticed and corrected. But, I'm not entirely sure. @BenHenning, @wxyxinyu, what do you think?

maitbayev replied:
What about including it in CodeRepl.py? It could be "rules" member variable. 
From my understanding, it needs to easy access for the backend/frontend.

seanlip replied:
Same issue; Karma tests won't be able to access it if that were the case.

maitbayev replied:
mock it ?

seanlip replied:
Sorry, I don't understand. Do you mean create a separate copy? I don't want to do that, I want the Karma tests to check against the exact rules, so that we can be sure that the frontend service definitions match the JSON file definitions exactly.

bigblind replied:
Maybe the [static-module](https://www.npmjs.com/package/static-module) npm package is something to look into here. It allows you to define modules of functions that are evaluated statically at build time. 

So for instance, you could define a `config` module, with a `getOption` function, and use it like this:

    var config = require("config");
    $(document.body).html("<h1>" + config.getOption("applicationTitle")+"</h1>");

And so after the build script has run, this gets converted into:

    var config = require("config");
    $(document.body).html("<h1>" +"Oppia"+"</h1>");

I have a browserify transforming function that I use for this in one of my previous projects, so maybe it's useful if I port that over. It means that only the necessary values or copied to the frontend, though it might increase the file size if config values are large, or repeated a lot.

seanlip replied:
Hm, I think we currently achieve the same thing (in this commit) using `karma-json-fixtures-preprocessor`. It allows the karma tests to read json files from the filesystem -- see my changes to `core/tests/karma.conf.js` and `extensions/interactions/rulesSpec.js` in this commit. If I'm not mistaken, static-module does basically the same thing, right?

Also, @maitbayev, FYI, the rulesSpec.js file contains the tests I was referring to in my reply -- they check that the code conforms to the single source of truth, which is intended to be the rules.json file.

BenHenning replied:
I actually think I lean toward rules belonging as part of their definition of the interaction used by them. Although it's convenient to put it in a rules.json file, we're sort of in this awkward midway between having rules separate from interactions and having them part of the definition. From what I understand, we'd rather have interactions be self-contained (especially since rules and objects are the only parts of an interaction not defined directly within it). I don't think we should skimp on this particular effort, if it's not too unreasonable to put them within the respective interaction definitions.

Unless I'm mistaken, we really only need rule descriptions for validating explorations. The only other locations it exists (backend classification and backend rule tests) are/can both being/be moved to the frontend. I think this leads to us to requiring roughly one semi-complicated function for retrieving these descriptions in an effort to validate explorations. I think that's worthwhile.

seanlip replied:
One sec -- are you talking here about putting the rule descriptions etc. in the interaction Python files, or creating a separate rules.json for each interaction? The former won't work: I have a test in the frontend (see rulesSpec.js) that checks that each rules service in the frontend corresponds exactly to the rules defined in the JSON file. Testing this requires that the frontend be able to access the rule definitions without the App Engine server running, so putting the definitions in the Python file will not work.

Also, @wxyxinyu and I discussed this and we realized that we can have a separate rules.json file for each interaction after all. The only thing that's needed is a canonical list of interactions whose validity we should be checking, so we plan to pull out feconf.ALLOWED_INTERACTION_CATEGORIES into a JSON file as well, and use that in the Karma test to construct the URLs for the individual rules.json files.

However, I think I should check that this sounds good before going ahead and doing it. What are your thoughts?

BenHenning replied:
Yeah, my response was hinging on details I didn't quite understand. :) Having a separate rules.json file per interaction seems just as reasonable, so yep I like it.

BenHenning commented:
(I haven't reviewed this yet, I'm respond to initial questions). @seanlip and I discussed some of this on Monday for his clarification. In particular regards to the purpose of soft rules and future classifiers. There have been several evolutions since soft rules were originally introduced. Initially, soft rules were placeholders for implementing statistical classifiers (such as string_classifier) into the rule framework. Later, @stephenchiang's work led to the string statistical classifier not being able to be part of the rule framework, which then led to soft rules being redefined. At that point, we found that soft rules were actually a useful "backup" to the statistical classifier to ensure any answers submitted for classification which are part of the training data of an answer group would always be matched to that group (this is not guaranteed with a statistical classifier alone).

During the work to migrate answers to a new data model, this entire concept further evolved. The central classify functions have been upgraded in there to include the concept of *classification categorization*, which is an indication as to whether an answer has been classified to a hard rule, soft rule, statistical classifier, or the default outcome of an interaction. It has since evolved one more time after Sean and I discussed it. Now, soft rules are going to be removed entirely and are going to be implemented directly as part of reader.classify (and the frontend port of it). This leads to all existing rules being 'hard rules'. We are dropping this differentiated terminology, but it will still exist for answer categorization (ie, whether an answer was classified according to an explicit hard rule, an implicit soft classification, statistical classification, or has been matched to the default outcome). Hereafter, soft rules will be defined through an interaction instead of being actual rules.

To address your concern about why the string classifier is a top-level classification function, it is because we do not yet have a generalized statistical classification framework. When we implement another statistical classifier, we will likely implement a statistical classifier registry and generalize the current function in reader (or rather, the equivalent in the frontend classification service). Which statistical classifier is associated with an interaction will also be defined as part of the interaction.

Finally, regarding removing 'submit_answer' from test_utils.py, I am actually doing this as part of my work with the answer migration. It's part of my overall strategy to cleanup the rule pipeline and I think it's the last major part to remove. The tests in editor_test and reader_test (unless the latter have been removed) should be using more isolated means of submitting answers, rather than depending on the classification process. Since the major one here is the training interface, an end-to-end test will need to be added to ensure it works across the frontend/backend boundary and properly interacts with the classification process properly.

BenHenning commented on core/controllers/editor_test.py:
```diff
@@ -175,34 +190,40 @@ def _put_and_expect_400_error(payload):
         self.logout()
 
     def test_resolved_answers_handler(self):
```
FYI: This test has been removed in the answer migrations branch, since resolved answers are obsolete.

BenHenning commented on core/controllers/reader.py:
```diff
@@ -100,21 +100,45 @@ def test_can_play(self, exploration_id, **kwargs):
+        normalized_param = obj_cls.normalize(parsed_param)
+        param_list.append(normalized_param)
+
+    # NOTE TO DEVELOPERS: These lines will not work -- they originally
```
This function essentially needs to be removed, along with all of the classify functions below. The only backend dependency on backend classification now is submit_answer, which will be removed as part of the answer migrations work.

BenHenning commented on core/controllers/reader.py:
```diff
@@ -133,7 +157,7 @@ def classify_hard_rule(state, params, input_type, normalized_answer, fs):
 
 
-def classify_soft_rule(state, params, input_type, normalized_answer, fs):
+def classify_soft_rule(state, params, input_type, normalized_answer):
```
A note to developers here: classify_soft_rules() is going to completely change as part of the answer migration (it will no longer call any rule evaluations).

BenHenning commented on core/controllers/reader_test.py:
```diff
@@ -109,104 +109,6 @@ def test_published_explorations_are_visible_to_logged_in_users(self):
         self.assertEqual(response.status_int, 200)
 
 
-class ReaderControllerEndToEndTests(test_utils.GenericTestBase):
```
Have these tests been replaced with end-to-end tests?

BenHenning replied:
In retrospect and in lieu of our recent discussions on e2e tests, I'm thinking these should actually be frontend karma tests. Thoughts?

BenHenning commented on core/domain/exp_domain.py:
```diff
@@ -519,7 +519,7 @@ def __init__(self, outcome, rule_specs):
         self.outcome = outcome
 
-    def validate(self, obj_type, exp_param_specs_dict):
+    def validate(self, interaction_rules, exp_param_specs_dict):
```
This is so much cleaner. :+1:

BenHenning commented on core/domain/exp_domain.py:
```diff
@@ -534,25 +534,21 @@ def validate(self, obj_type, exp_param_specs_dict):
             rule_spec.validate(
-                rule_domain.get_param_list(rule_class.description),
+                rule_domain.get_param_list(
+                    interaction_rules[rule_spec.rule_type]['description']),
```
Probably should use ``get_rule_description()`` here for consistency (and, hence, pass in interaction into this validate() instead).

seanlip replied:
Good idea; done.

BenHenning commented on core/domain/rule_domain.py:
```diff
@@ -16,65 +16,13 @@
 
 """Classes relating to rules."""
```
Is there an advantage to keeping rule_domain around? I feel as though we should just get rid of it.

seanlip replied:
Done.

BenHenning commented on core/domain/exp_domain.py:
```diff
@@ -1145,8 +1138,17 @@ def update_interaction_answer_groups(self, answer_groups_list):
+                    param_list = rule_domain.get_param_list(
+                        matched_rule_description)
+
+                    param_type = None
```
Since ``get_param_list()`` is only used in rule_domain_test and exp_domain and rule_domain* will probably be removed, we might as well just implement a ``get_param_type_from_rule_description()`` function which combines this logic with the existing logic in ``get_param_list()``. It keeps this code cleaner and avoids having a strange function which we only partially using.

BenHenning replied:
Later note: I think we should add this new function within interaction, that way interaction is the single source of truth for all rule information in the backend (which will be consistent with our vision for the frontend interaction definitions).

seanlip replied:
Good idea -- done. I added two new functions to extensions/interactions/base.py and removed rule_domain.

BenHenning commented on extensions/interactions/base_test.py:
```diff
@@ -150,6 +148,32 @@ def test_interaction_properties(self):
+                interaction_id)
+            self.assertEqual(len(interaction.rules), expected_num)
+
+        _check_num_interaction_rules('MultipleChoiceInput', 1)
```
What about the other interactions?

seanlip replied:
I think this is more of a sanity check on the registry and the .rules property than a check on individual interactions, so, leaving as-is.

BenHenning commented on extensions/interactions/base_test.py:
```diff
@@ -150,6 +148,32 @@ def test_interaction_properties(self):
+        with self.assertRaises(KeyError):
+            _check_num_interaction_rules('FakeObjType', 0)
+
+    def test_interaction_rule_descriptions(self):
```
Nit: Please rename to ``test_interaction_rule_descriptions_in_dict`` to indicate this is not testing ``get_rule_description()``.

BenHenning replied:
Also, can we move these (and related) tests to the frontend, instead? Seems a bit awkward to validate rules in the backend when they are fully defined in the frontend.

seanlip replied:
Fixed the nit.

I don't think so; this is really testing the generation of the interaction dict, the logic for which is in the backend.

BenHenning replied:
Hmm, I still have mixed opinions about it. The backend really doesn't use that portion of the interaction dict much, anyway. Either way, I think I'm fine with it staying where it is. I still think it's slightly awkward, though.

BenHenning commented on extensions/interactions/base_test.py:
```diff
@@ -313,6 +337,28 @@ def test_default_interactions_are_valid(self):
+                param_list = rule_domain.get_param_list(rule_description)
+
+                for (_, param_obj_type) in param_list:
+                    # TODO(sll): Get rid of these special cases.
```
Why do we need special cases? If any objects don't have object editor templates, can they be removed?

seanlip replied:
Yes, the only reason this is here is the fuzzy-rules issue, I think (and perhaps nonnegativeint, since special-case logic in the frontend changes that to a dropdown of choices for multiple choice). @wxyxinyu -- if we end up removing the fuzzy rules in this commit, could you get rid of the relevant object types here and in extensions.objects too?

BenHenning commented:
Thanks @seanlip! I had several comments. I also some in-line pieces of documentation for either your reference or another developer's if you don't end up finishing this up yourself.

seanlip commented on core/controllers/editor_test.py:
```diff
@@ -266,38 +275,50 @@ def _create_training_data(*arg):
             self.assertEqual(
                 response_dict['unhandled_answers'],
-                _create_training_data('joyful', 'sad'))
+                _create_training_data('sad', 'joyful'))
```
NB: @BenHenning, I had to switch the order here (after using event_services.AnswerSubmissionEventHandler.record() instead), otherwise an error would occur. However, I'm confused about why I needed to do this. I'm under the impression that order doesn't really matter here, but do you think it needs a more careful look?

/cc @wxyxinyu

BenHenning replied:
Yes, at initial glance this makes me slightly uncomfortable. The order for the response training data is very specific (answers are returned based on frequency). The tiebreaker is not clear to me, but I think it's deterministic and based on insertion order. Since insertion order didn't change, I'm slightly confused as well. @wxyxinyu when you address this, I can help you look into it more if you'd like. It might also not be an issue since this is largely based on a nearly deprecated answers storage system.

seanlip commented:
Thanks, @BenHenning! I addressed some comments (and replied to some others). Thanks also for the top-level comment; I've implemented that, and it's now only the checks in reader_test that fail.

I've also discussed this PR with @wxyxinyu, and am going to hand over the remainder of the migration to her.

BenHenning commented:
The follow-up commit looks good to me, as well. I'll do a complete pass again once the remaining changes and migration is complete. (This being @seanlip or @wxyxinyu depending on who finishes the PR).

seanlip commented:
Update from conversation with @wxyxinyu: we'll probably leave the json migration stuff (i.e. migrating feconf.ALLOWED_INTERACTION_CATEGORIES to a json file and using one rules.json file per interaction) for a separate commit/PR, just to keep this one manageable. The immediate thing to do is to just migrate the rule evaluation/classification to the frontend (deprecating soft rules in the process), and check that in.

seanlip commented:
Update from conversation with @wxyxinyu: we'll probably leave the json migration stuff (i.e. migrating feconf.ALLOWED_INTERACTION_CATEGORIES to a json file and using one rules.json file per interaction) for a separate commit/PR, just to keep this one manageable. The immediate thing to do is to just migrate the rule evaluation/classification to the frontend (deprecating soft rules in the process), and check that in.

wxyxinyu commented:
Hi @seanlip, @BenHenning, I've made some updates, PTAL.

Further work required:
[ ] add more frontend tests for classify
[ ] add e2e tests
First one is probably needed before merge, second one is not.

seanlip commented on core/controllers/editor_test.py:
```diff
@@ -175,34 +175,42 @@ def _put_and_expect_400_error(payload):
-        result_dict = self.submit_answer('0', state_name, '0')
+        first_state_name = exploration_dict['exploration']['init_state_name']
+        event_services.AnswerSubmissionEventHandler.record(
+            exp_id, 1, first_state_name, 'N/A', 0)
```
Is "N/A" a valid rulespec string?

seanlip commented on core/domain/exp_domain.py:
```diff
@@ -493,9 +492,9 @@ def validate(self):
-    rules are ORed together. Answer groups may also support fuzzy/implicit
-    rules that involve soft matching of answers to a set of training data
-    and/or example answers dictated by the creator.
+    rules are ORed together. Answer groups may also support a classifier/
```
Do you mean:

    ...may also support a classifier that performs soft matching of answers to ...

wxyxinyu replied:
fixed

seanlip commented on core/domain/exp_domain.py:
```diff
@@ -533,35 +532,30 @@ def validate(self, obj_type, exp_param_specs_dict):
-        """Will return the answer group's fuzzy rule index, or None if it
-        doesn't exist.
+    def get_classifier_index(self):
+        """Returns the index of the classifier in the answer group's, or None
```
Remove apostrophe-s on "answer group".

wxyxinyu replied:
done

seanlip commented on core/domain/obj_services_test.py:
```diff
@@ -15,6 +15,7 @@
 """Tests for services relating to typed objects."""
 
 from core.domain import obj_services
+from core.domain import interaction_registry
```
nit: alphabetize

wxyxinyu replied:
done

seanlip commented on core/domain/obj_services_test.py:
```diff
@@ -63,3 +64,43 @@ def test_object_js_filenames(self):
+        object_default_vals = obj_services.get_default_object_values()
+
+        for interaction in interactions:
+            for rule_name in interaction.rules.keys():
```
drop .keys()

wxyxinyu replied:
done

seanlip commented on core/domain/obj_services_test.py:
```diff
@@ -63,3 +64,43 @@ def test_object_js_filenames(self):
+                param_list = interaction.get_rule_param_list(rule_name)
+
+                for (_, param_obj_type) in param_list:
+                    param_name = param_obj_type.__name__
```
Should param_name be param_type or param_type_name? Otherwise there's potential for confusion with the name of the parameter instance.

wxyxinyu replied:
ok

seanlip commented on core/templates/dev/head/player/AnswerClassificationService.js:
```diff
@@ -17,19 +17,19 @@
+  '$http', '$q', 'LearnerParamsService', 'INTERACTION_SPECS',
+  'ENABLE_STRING_CLASSIFIER',
+  function($http, $q, LearnerParamsService, INTERACTION_SPECS,
+      ENABLE_STRING_CLASSIFIER, CLASSIFIER_RULE_TYPE) {
```
Oh oh ... function args do not match stringified dependencies.

wxyxinyu replied:
done

seanlip commented on core/templates/dev/head/player/AnswerClassificationService.js:
```diff
@@ -46,8 +46,9 @@ oppia.factory('answerClassificationService', [
           var ruleSpec = answerGroups[i].rule_specs[j];
-          if (interactionRulesService[ruleSpec.rule_type](
-              answer, ruleSpec.inputs)) {
+          if (ruleSpec.rule_type != CLASSIFIER_RULE_TYPE &&
```
If you're not writing e2e tests in this PR, at least run through the workflow manually to make sure everything works and that there aren't any console errors, otherwise it's possible to miss things like the dependency issue indicated above.

BenHenning replied:
Please add Karma tests for this service. Most of the backend tests that were deleted in this PR should be portable to Karma to ensure the same behavior stays consistent, if there aren't already sufficient tests. Sean's right, this line should have broken at least 1 Karma test.

seanlip commented on core/templates/dev/head/player/AnswerClassificationService.js:
```diff
@@ -81,8 +82,7 @@ oppia.factory('answerClassificationService', [
        * @param {boolean} isInEditorMode - Whether the function is being called
        *   in editor mode.
        * @param {?function} interactionRulesService - The service which contains
```
Is the question mark still needed?

wxyxinyu replied:
nope, it's all migrated :D

seanlip commented on core/templates/dev/head/player/AnswerClassificationService.js:
```diff
@@ -81,8 +82,7 @@ oppia.factory('answerClassificationService', [
        * @param {?function} interactionRulesService - The service which contains
-       *   the rules of that interaction. If this is undefined, then the
-       *   function uses server-side classification.
+       *   the rules of that interaction.
```
Just to check, does this refer to 'hard' rules only? If so, might be worth adding the word 'hard' here.

BenHenning replied:
Nit: please refer to 'hard' rules as explicit rules instead, as that's the new terminology adopted as part of the answer migration work.

wxyxinyu replied:
done

seanlip commented on core/templates/dev/head/player/AnswerClassificationService.js:
```diff
@@ -97,13 +97,20 @@ oppia.factory('answerClassificationService', [
+        var answerGroups = oldState.interaction.answer_groups;
+        var defaultOutcome = oldState.interaction.default_outcome;
+
         if (interactionRulesService) {
```
Is there a case when there's no interactionRulesService?

wxyxinyu replied:
might be better to add a check that it does exist. this might prevent errors when people are adding new interactions.

seanlip replied:
SGTM, good idea

seanlip commented on core/templates/dev/head/player/AnswerClassificationService.js:
```diff
@@ -97,13 +97,20 @@ oppia.factory('answerClassificationService', [
-          deferred.resolve(classifyAnswer(
-            answer, answerGroups, defaultOutcome, interactionRulesService));
-        } else {
+        if (result === null || (result.outcome == defaultOutcome &&
```
See previous comment -- you _might_ not need the null check here.

seanlip commented on core/templates/dev/head/base.js:
```diff
@@ -25,10 +25,13 @@ oppia.constant('PLACEHOLDER_OUTCOME_DEST', '/');
-oppia.constant('DEFAULT_FUZZY_RULE', {
-  rule_type: 'FuzzyMatches',
+// Feature still in development. Synchronize this with feconf.
+oppia.constant('ENABLE_STRING_CLASSIFIER', false);
```
Add test to make sure it's synced?

seanlip commented on core/templates/dev/head/player/AnswerClassificationService.js:
```diff
@@ -97,13 +97,20 @@ oppia.factory('answerClassificationService', [
-        } else {
+        if (result === null || (result.outcome == defaultOutcome &&
+            INTERACTION_SPECS[oldState.interaction.id]
+              .is_string_classifier_trainable &&
```
Might be nicer to declare `oldInteraction = INTERACTION_SPECS[oldState.interaction.id]` as a separate variable in order to avoid requiring a newline before the dot.

wxyxinyu replied:
ok

seanlip commented on core/tests/karma.conf.js:
```diff
@@ -26,7 +26,9 @@ module.exports = function(config) {
       'extensions/**/*.js',
-      'extensions/interactions/**/*.html'
+      'extensions/interactions/**/*.html',
+      'extensions/skins/**/*.html',
```
This line is not needed? We've removed skins...

wxyxinyu replied:
removed

seanlip commented on core/tests/karma.conf.js:
```diff
@@ -68,7 +70,9 @@ module.exports = function(config) {
       'core/templates/dev/head/components/rating_display.html': ['ng-html2js'],
-      'extensions/interactions/**/*.html': ['ng-html2js']
+      'extensions/interactions/**/*.html': ['ng-html2js'],
+      'extensions/skins/**/*.html': ['ng-html2js'],
```
Ditto.

wxyxinyu replied:
done

seanlip commented on extensions/interactions/base.py:
```diff
@@ -143,6 +143,25 @@ def _stats_log_template(self):
             return '{{answer}}'
 
     @property
+    def rules(self):
```
Thinking about it, should this be renamed rules_dict? rules suggests a domain object...

wxyxinyu replied:
Not really sure I agree with that. The plural 'rules' does seem like a list of something to me.

seanlip replied:
That's reasonable, although I think it actually supports the suggestion I made (but for a different reason), since you're returning a dict here and not a list. If 'rules' suggests a list then the method name is problematic.

wxyxinyu replied:
done

seanlip commented on extensions/interactions/base.py:
```diff
@@ -192,23 +211,48 @@ def to_dict(self):
+        """Gets the parameter type for a given rule parameter name."""
+        rule_param_list = self.get_rule_param_list(rule_name)
+
+        for item in rule_param_list:
```
Can this be "for (..., ...) in rule_param_list", with each of the ellipsis replaced by a meaningful noun?

wxyxinyu replied:
done

seanlip commented on extensions/interactions/base_test.py:
```diff
@@ -380,3 +415,6 @@ def test_linear_interactions(self):
         self.assertEqual(count, 1)
+
+    def test_objects_in_rules_have_default_values(self):
+        pass
```
?

wxyxinyu replied:
I think I forgot to write that test :P

seanlip replied:
:) Are you going to write it?

seanlip replied:
Done -- made it a part of test_default_interactions_are_valid.

seanlip commented on feconf.py:
```diff
@@ -415,6 +421,7 @@ def get_empty_ratings():
 
 # Unfinished features.
 SHOW_TRAINABLE_UNRESOLVED_ANSWERS = False
+# Synchronize this with base.js
```
Reads a bit like a TODO. Maybe: "NOTE TO DEVELOPERS: This should be synchronized with the value in base.js."?

wxyxinyu replied:
done

seanlip commented on core/templates/dev/head/base.js:
```diff
@@ -25,10 +25,13 @@ oppia.constant('PLACEHOLDER_OUTCOME_DEST', '/');
 
-oppia.constant('DEFAULT_FUZZY_RULE', {
-  rule_type: 'FuzzyMatches',
+// Feature still in development. Synchronize this with feconf.
```
Reads a bit like a TODO. Maybe:

    Feature still in development.
    NOTE TO DEVELOPERS: This should be synchronized with the value in feconf.

wxyxinyu replied:
done

seanlip commented:
This is _amazing_. Thanks, @wxyxinyu! I just took a pass, and only have minor comments. I'm very excited about how much code is being deleted!

I agree that this needs some more unit tests prior to merge. It might also be worth just speccing out the e2e ones as well, since you have a good idea of what the gaps are. Maybe file a new issue outlining what such test(s) should do?

Thanks!

BenHenning commented on core/controllers/reader.py:
```diff
@@ -72,97 +70,29 @@ def test_can_play(self, exploration_id, **kwargs):
-        fuzzy_rule_spec_index = answer_group.get_fuzzy_rule_index()
-        if fuzzy_rule_spec_index is not None:
-            fuzzy_rule_spec = answer_group.rule_specs[fuzzy_rule_spec_index]
+        classifier_spec_index = answer_group.get_classifier_index()
```
I'm not so sure about this naming. Fuzzy rules are not the same thing as classifiers. They are still technically rules which contain training data used to build statistical classifiers. This might need more thought that goes outside the scope of this PR. I'm not currently recalling what's supposed to happen with fuzzy rules.

BenHenning commented on core/controllers/reader.py:
```diff
@@ -218,18 +148,13 @@ def classify(exp_id, state, answer, params):
-            feconf.ENABLE_STRING_CLASSIFIER and response is None):
+            feconf.ENABLE_STRING_CLASSIFIER):
         response = classify_string_classifier_rule(state, normalized_answer)
+    elif feconf.ENABLE_STRING_CLASSIFIER:
```
Shouldn't this be ``else:``?

wxyxinyu replied:
if the string classifier does not exist we only want to raise an exception when the ENABLE_STRING_CLASSIFIER flag is set to true, and all interactions are associated with a classifer.

seanlip replied:
When ENABLE_STRING_CLASSIFIER is False, will this handler be called anyway?

wxyxinyu replied:
hmm yeah good point. I think we can actually get rid of ENABLE_STRING_CLASSIFIER in the backend, since the frontend checks.

or should we still do a backend check?

seanlip replied:
How about an overall top-level backend check at the start of the handler? And then assume, after then, that it's set.

wxyxinyu replied:
done

BenHenning commented on core/controllers/reader.py:
```diff
@@ -194,7 +124,7 @@ def classify_string_classifier_rule(state, normalized_answer):
 # those rules is picked.
-def classify(exp_id, state, answer, params):
+def classify(state, answer):
     """Normalize the answer and select among the answer groups the group in
```
Please update this docstring. We are redefining classify to only deal with statistical classification, so everywhere along its call path should be updated accordingly.

wxyxinyu replied:
done

BenHenning commented on core/controllers/reader.py:
```diff
@@ -194,7 +124,7 @@ def classify_string_classifier_rule(state, normalized_answer):
 # TODO(bhenning): Add more tests for classification, such as testing multiple
```
Should this TODO still be here? This PR should introduce frontend tests which satisfy this TODO.

wxyxinyu replied:
removed the TODO, but please check the tests.

BenHenning commented on core/controllers/reader.py:
```diff
@@ -456,7 +381,7 @@ class ClassifyHandler(base.BaseHandler):
 
     @require_playable
-    def post(self, exploration_id):
+    def post(self, unused_exploration_id):
```
Is there any reason to keep ``unused_exploration_id`` around? I think the answer is yes (for the future classifier work), in which case please add a TODO to explain why there is an unused parameter here and what it will be used for in the future.

wxyxinyu replied:
oh, it's actually because @require_playable needs that field. otherwise the code gets a bit ugly replicating the @require_playable. It's also possible this will be used for future classifier work. added note.

BenHenning commented on core/domain/exp_domain.py:
```diff
@@ -354,7 +353,7 @@ def __init__(self, rule_type, inputs):
     def stringify_classified_rule(self):
         """Returns a string representation of a rule (for the stats log)."""
-        if self.rule_type == rule_domain.FUZZY_RULE_TYPE:
+        if self.rule_type == feconf.CLASSIFIER_RULE_TYPE:
```
Prefer moving CLASSIFIER_RULE_TYPE to exp_domain. I think we want to move away from using feconf as a general bucket for constants. I think rule-related constants appropriately fit within exp_domain since nothing left in the backend really cares about rules.

wxyxinyu replied:
done

BenHenning commented on core/domain/exp_domain.py:
```diff
@@ -533,35 +532,30 @@ def validate(self, obj_type, exp_param_specs_dict):
 
-        all_rule_classes = rule_domain.get_rules_for_obj_type(obj_type)
-        seen_fuzzy_rule = False
+        seen_classifier = False
```
Perhaps we should instead refer to ``classifier_rule`` here and in reader.py, to be consistent with the name of the method. I think that's a clearer name than just ``classifier``, since the latter is ambiguous with statistical classifier over a rule which contains training data for a statistical classifier.

BenHenning commented on core/domain/exp_domain.py:
```diff
@@ -533,35 +532,30 @@ def validate(self, obj_type, exp_param_specs_dict):
-                    r for r in all_rule_classes
-                    if r.__name__ == rule_spec.rule_type)
-            except StopIteration:
+            if rule_spec.rule_type not in interaction.rules:
```
:)

BenHenning commented on core/domain/exp_domain.py:
```diff
@@ -533,35 +532,30 @@ def validate(self, obj_type, exp_param_specs_dict):
                     raise utils.ValidationError(
-                        'AnswerGroups can only have one fuzzy rule.')
-                seen_fuzzy_rule = True
+                        'AnswerGroups can only have one classifier.')
```
s/classifier/classifier rule

BenHenning commented on core/domain/exp_domain.py:
```diff
@@ -533,35 +532,30 @@ def validate(self, obj_type, exp_param_specs_dict):
-    def get_fuzzy_rule_index(self):
-        """Will return the answer group's fuzzy rule index, or None if it
-        doesn't exist.
+    def get_classifier_index(self):
```
Here and elsewhere in this file: s/classifier/classifier rule

wxyxinyu replied:
done

BenHenning commented on core/domain/exp_domain_test.py:
```diff
@@ -346,18 +346,18 @@ def test_validation(self):
 
         self._assert_validation_error(
-            exploration, 'AnswerGroups can only have one fuzzy rule.')
+            exploration, 'AnswerGroups can only have one classifier.')
```
classifier rule?

wxyxinyu replied:
done

BenHenning commented on core/domain/obj_services.py:
```diff
@@ -71,3 +74,10 @@ def get_all_object_editor_js_templates():
+def get_default_object_values():
+    """Returns a dictionary containing the default object values."""
+
+    return json.loads(
```
Should this be cached? How often is this going to be hit? It seems like a lot for loading a file everytime.

wxyxinyu replied:
making this a TODO for now

BenHenning commented on core/domain/obj_services_test.py:
```diff
@@ -63,3 +64,43 @@ def test_object_js_filenames(self):
+    """Test that the default value of objects recorded in
+    extensions/interactions/object_defaults.json correspond to
+    the defined default values in objects.py for all objects that
+    are used in rules."""
```
Nit: ``"""`` should be on the next line by itself.

wxyxinyu replied:
done

seanlip replied:
Not done?

seanlip replied:
Now done

BenHenning commented on core/domain/obj_services_test.py:
```diff
@@ -63,3 +64,43 @@ def test_object_js_filenames(self):
+
+    def test_all_rule_input_fields_have_default_values(self):
+        """Checks that all rule input fields have a default value, and this
+        is provided in get_default_values()."""
```
Ditto for this docstring (please make sure ``"""`` is on its own line and does not have whitespace between it and the first statement.

wxyxinyu replied:
done

seanlip replied:
Not done?

seanlip replied:
Done

BenHenning commented on core/domain/obj_services_test.py:
```diff
@@ -63,3 +64,43 @@ def test_object_js_filenames(self):
+
+    def test_get_object_default_values_is_valid(self):
+        """Checks that the default values provided by get_default_values()
+        correspond to the ones defined in objects.py"""
```
Ditto for this docstring.

wxyxinyu replied:
done

BenHenning commented on core/domain/obj_services_test.py:
```diff
@@ -63,3 +64,43 @@ def test_object_js_filenames(self):
+                            param_name))
+                    self.assertIn(param_name, object_default_vals)
+                    self.assertEqual(default_value,
+                                     object_default_vals[param_name])
```
Prefer indenting on the next line, e.g.:

```python
self.assertEqual(
    default_value, object_default_vals[param_name])
```

wxyxinyu replied:
done

BenHenning commented on core/domain/obj_services_test.py:
```diff
@@ -63,3 +64,43 @@ def test_object_js_filenames(self):
+        all_object_classes = obj_services.Registry.get_all_object_classes()
+        for (obj_type, default_value) in object_default_vals.iteritems():
+            self.assertIn(obj_type, all_object_classes)
+            self.assertEqual(default_value,
```
Ditto for indentation.

wxyxinyu replied:
done

BenHenning commented on core/templates/dev/head/base.js:
```diff
@@ -25,10 +25,13 @@ oppia.constant('PLACEHOLDER_OUTCOME_DEST', '/');
+// Feature still in development. Synchronize this with feconf.
+oppia.constant('ENABLE_STRING_CLASSIFIER', false);
+
+oppia.constant('DEFAULT_CLASSIFIER', {
```
What about: ``DEFAULT_CLASSIFIER_RULE_SPEC``?

wxyxinyu replied:
done

BenHenning commented on core/templates/dev/head/components/AnswerGroupEditorDirective.js:
```diff
@@ -157,14 +157,14 @@ oppia.directive('answerGroupEditor', [function() {
           if (!ruleType) {
             // This should never happen. An interaction must have more than just
-            // a fuzzy rule, as verified in a backend test suite:
+            // a classifier, as verified in a backend test suite:
```
Here and elsewhere: s/classifier/classifier rule

wxyxinyu replied:
done

seanlip replied:
not done?

seanlip replied:
done

BenHenning commented on core/templates/dev/head/components/ClassifierPanelDirective.js:
```diff
@@ -13,17 +13,17 @@
  */
 
-oppia.directive('fuzzyRulePanel', [function() {
+oppia.directive('classifierPanel', [function() {
```
classifierRulePanel?

wxyxinyu replied:
done

BenHenning commented on core/templates/dev/head/editor/StateEditor.js:
```diff
@@ -336,27 +336,27 @@ oppia.factory('trainingDataService', [
-        // Ensure the answer group has a fuzzy rule.
-        var fuzzyRule = null;
+        // Ensure the answer group has a classifier.
+        var classifier = null;
```
s/classifier/classifierRule

seanlip replied:
done

BenHenning commented on core/templates/dev/head/player/AnswerClassificationService.js:
```diff
@@ -17,19 +17,19 @@
  */
 
 oppia.factory('answerClassificationService', [
```
Should be ``AnswerClassificationService``

wxyxinyu replied:
done

BenHenning commented on core/templates/dev/head/player/AnswerClassificationService.js:
```diff
@@ -97,13 +97,20 @@ oppia.factory('answerClassificationService', [
+            INTERACTION_SPECS[oldState.interaction.id]
+              .is_string_classifier_trainable &&
+            ENABLE_STRING_CLASSIFIER)) {
           // TODO(bhenning): Figure out a long-term solution for determining
```
I think we can safely say this TODO is deprecated. :) Please remove.

BenHenning commented on core/templates/dev/head/player/AnswerClassificationServiceSpec.js:
```diff
@@ -34,6 +44,7 @@ describe('Answer classification service', function() {
         value: 'content'
       }],
       interaction: {
+        id: 'RuleTest',
```
Please make sure there are tests in this spec that result in needing to call the backend classification service.

seanlip replied:
I believe this is now handled -- see the "should query the backend if no answer group matches" test below.

BenHenning commented on core/tests/protractor.conf.js:
```diff
@@ -135,7 +135,7 @@ exports.config = {
         baseDirectory: '../protractor-screenshots',
         // Function to build filenames of screenshots
-        pathBuilder: function pathBuilder(spec, descriptions, results, capabilities) {
+        pathBuilder: function(spec, descriptions, results, capabilities) {
```
I'm amazed the previous JS didn't fail...good find.

BenHenning commented on data/explorations/classifier_demo_exploration.yaml:
```diff
@@ -0,0 +1,76 @@
+author_notes: ''
```
Shouldn't this file be ``git mv``'d?

wxyxinyu replied:
Probably. But changing the history at this point seems a lot of trouble. Is it ok if I just left it?

seanlip replied:
FWIW, I think it's fine.

BenHenning commented on extensions/interactions/base.py:
```diff
@@ -78,7 +79,7 @@ class BaseInteraction(object):
     # Whether the interaction has only one possible answer.
     is_linear = False
-    # Whether this interaction supports training and fuzzy classification.
+    # Whether this interaction supports training and classification.
```
Nit: prefer ``statistical classification`` to be clear.

wxyxinyu replied:
done

BenHenning commented on extensions/interactions/base.py:
```diff
@@ -143,6 +143,25 @@ def _stats_log_template(self):
+        if self._cached_rules is not None:
+            return self._cached_rules
+
+        rules_index_dict = json.loads(
```
Can we load the ``rules.json`` file somewhere else 1 time? It might be more performance benefiting to cache the JSON file in 1 location, then just lookup the subdict for this property instead of loading the JSON file for each interaction type. WDYT?

wxyxinyu replied:
Probably, but there is also a TODO to split up the json file into individual interactions, so I think worrying about caching after that is probably best.

BenHenning commented on extensions/interactions/base.py:
```diff
@@ -192,23 +211,48 @@ def to_dict(self):
-        except StopIteration:
-            raise Exception('Could not find rule with name %s' % rule_name)
+            bar_index = description.find('|')
+            param_name = description[: bar_index]
```
Nit: No space after ``:``

wxyxinyu replied:
done

BenHenning commented on extensions/interactions/base.py:
```diff
@@ -192,23 +211,48 @@ def to_dict(self):
+            description = description[bar_index + 1:]
+
+            closing_index = description.find('}}')
+            normalizer_string = description[: closing_index]
```
Ditto for this ``:``

wxyxinyu replied:
done

BenHenning commented on extensions/interactions/rulesSpec.js:
```diff
@@ -0,0 +1,53 @@
+// Copyright 2014 The Oppia Authors. All Rights Reserved.
```
2016? :)

BenHenning commented on extensions/interactions/rulesSpec.js:
```diff
@@ -0,0 +1,53 @@
+describe('Rule spec services', function() {
+  var ruleSpecsServices = {};
+  var rulesJson;
+  var IMPLICIT_RULE_NAME = 'ClassifyMatches';
```
This really ought to be ``CLASSIFY_RULE_NAME`` to stay consistent with implementation code.

seanlip replied:
Agreed, or rather CLASSIFIER_RULESPEC_STR. @wxyxinyu can you fix this please?

seanlip replied:
Done

BenHenning commented on feconf.py:
```diff
@@ -42,14 +42,20 @@
+    'extensions', 'interactions', 'object_defaults.json')
+
+# Constants related to rules.
+CERTAIN_TRUE_VALUE = 1.0
```
I'm inclined to think these are no longer necessary. The notion of having levels of truth was deprecated after @stephenchiang implemented the string classifier. Can we get rid of these true and false variants?

wxyxinyu replied:
done

BenHenning commented on feconf.py:
```diff
@@ -42,14 +42,20 @@
+# Constants related to rules.
+CERTAIN_TRUE_VALUE = 1.0
+CERTAIN_FALSE_VALUE = 0.0
+CLASSIFIER_RULE_TYPE = 'ClassifyMatches'
```
In an earlier comment, I suggested moving this to exp_domain instead, to avoid continuing to spill constants into feconf.

wxyxinyu replied:
done

BenHenning commented on feconf.py:
```diff
@@ -42,14 +42,20 @@
     'core', 'templates', TEMPLATES_DIR_PREFIX, 'head')
 DEPENDENCIES_TEMPLATES_DIR = os.path.join('extensions', 'dependencies')
 VALUE_GENERATORS_DIR = os.path.join('extensions', 'value_generators')
+OBJECT_DEFAULT_VALUES_FILE_PATH = os.path.join(
```
Should there be a similar constant for ``rules.json`` as well, to avoid doing the ``os.path.join()`` work in interactions/base.py?

wxyxinyu replied:
should that constant be in base.py or here?

seanlip replied:
Let's put it here, with the other similar ones.

BenHenning commented on core/tests/karma.conf.js:
```diff
@@ -94,16 +98,21 @@ module.exports = function(config) {
       }
     },
     plugins: [
+      'karma-jasmine-jquery',
```
Why do we need this? It doesn't seem like anything extra is being done in Karma tests.

wxyxinyu replied:
seems to be a dependency somewhere

seanlip replied:
I remember adding it, it is for the json fixture I think

BenHenning commented:
Thanks @wxyxinyu! I have to concur with @seanlip, this really does look fantastic. Seeing this much code cleaned up is really, really awesome to see. Plus, we're data driving more of our codebase, which is equally fantastic.

A couple of things:

1. I didn't do quite as thorough of a review as Sean, though I did try and focus on anything that might conflict with other answer work going on. I don't think that's really an issue. There's some small overlap, but I'm not too concerned. This is surprisingly well-isolated from the parts of answers that I'm working on.
1. After reading through the code, I'm very inclined to suggest you instead rename ``fuzzy rule`` to be ``classifier rule`` instead of just ``classifier`` to avoid ambiguity with the actual classifier logic classes
1. In https://github.com/oppia/oppia/pull/1318/files#r63124950 I suggest implementing Karma tests instead of e2e; I'm inclined to think we don't need any new e2e tests as part of this PR
1. I'm worried that we are losing testing coverage. I didn't go through it with a fine-toothed comb, but can you double check to make sure all of the tests that were deleted have replacements in the frontend? It might even be better if there are also new tests added on top of those. This touches very critical components, so I want to be sure it's among the best tested code in the codebase.

Thanks @wxyxinyu!!! This is awesome work.

seanlip commented:
@BenHenning -- this should be ready for a review. Please could you take a look?

Also, @wxyxinyu has a question: could you give her some ideas about what other tests she should add, especially to replace those in reader_test?

Thanks!

seanlip commented on core/controllers/reader.py:
```diff
@@ -455,8 +371,12 @@ class ClassifyHandler(base.BaseHandler):
+        """Handle POST requests.
+
+        Note: unused_exploration_id is needed because @require_playable needs 2
+        arguments
```
nit: add full stop at end.

seanlip replied:
done

seanlip commented on core/controllers/reader_test.py:
```diff
@@ -252,79 +153,23 @@ def _get_classiying_rule_type(self, answer):
-            self._get_classiying_rule_type('I guessed.'),
-            'soft')
+        return (answer_group.get_classifier_rule_index() == rule_spec_index and
+                predict_counter.times_called > 0)
```
can we have an equality check here?

seanlip replied:
done

seanlip commented on core/domain/obj_services.py:
```diff
@@ -71,3 +74,11 @@ def get_all_object_editor_js_templates():
+
+def get_default_object_values():
+    """Returns a dictionary containing the default object values."""
+    #TODO(wxy): Cache this as it is accessed many times.
```
nit: add single space after `#`

seanlip replied:
done

seanlip commented on core/templates/dev/head/player/AnswerClassificationService.js:
```diff
@@ -97,13 +97,23 @@ oppia.factory('answerClassificationService', [
+            answer, answerGroups, defaultOutcome, interactionRulesService);
         } else {
+          alertsService.addWarning(
+            'Something went wrong with the exploration.');
```
Can this error be a little more explicit, like: "Something went wrong with the exploration: no interactionRulesService provided" (or something like that)? Otherwise, debugging things in the future is going to be trickier, since you have this same message in multiple locations.

seanlip replied:
Done

seanlip commented on core/templates/dev/head/player/AnswerClassificationServiceSpec.js:
```diff
@@ -75,30 +88,9 @@ describe('Answer classification service', function() {
-    $httpBackend.verifyNoOutstandingExpectation();
-    $httpBackend.verifyNoOutstandingRequest();
+  it('should be undefined if no frontend rules are provided', function() {
+    expect(acs.getMatchingClassificationResult(explorationId, state, 0, false))
```
Should this check that some alert should be thrown? In fact, should we throw an exception and break noisily since the learner won't be able to proceed?

wxyxinyu replied:
yep. changing it to reject the promise instead. returning different types of things depending on the case was probably also a really bad idea lol

seanlip commented on extensions/interactions/object_defaults.json:
```diff
@@ -0,0 +1,24 @@
+    "SetOfNormalizedString": [],
+    "SetOfUnicodeString": [],
+    "UnicodeString": ""
+}
```
add newline at EOF?

seanlip replied:
done

seanlip commented:
Took a lightweight pass. Had a few comments, but overall I think it looks great -- thanks @wxyxinyu!

(@BenHenning, I still recommend taking a pass, but mainly focusing on the statistical classifier aspects, since you're probably more familiar with that than I am. Thanks!)

seanlip commented:
Remaining comments that need addressing:
- https://github.com/oppia/oppia/pull/1318#discussion_r62817010
- https://github.com/oppia/oppia/pull/1318#discussion-diff-63124762
- https://github.com/oppia/oppia/pull/1318#discussion-diff-63124841
- https://github.com/oppia/oppia/pull/1318#discussion_r67253193

seanlip commented:
Is this mergeable? It looks fine to me...

wxyxinyu commented:
Added one last thing. I think this is mergeable when tests pass.
